File: day21/day21.kt
import java.io.File

fun main() {
    val input = File("day21.txt").readLines().map {
        val m = Regex("""(.+) \(contains (.+)\)""").matchEntire(it)!!
        Pair(m.groupValues[1].split(" ").toSet(), m.groupValues[2].split(", ").toSet())
    }

    val allergens = mutableMapOf<String,String>()
    val unknown = input.flatMap { it.second }.toMutableSet()
    while (unknown.isNotEmpty()) {
        val (l, r) = unknown.asSequence().mapNotNull { test ->
            val (l, r) = input.filter { it.second.contains(test) }.unzip().toList()
                .map { it.reduce { a, b -> a intersect b }.toMutableSet() }
            l -= allergens.keys
            r -= allergens.values
            if (l.isNotEmpty() && l.size == r.size) l to r else null
        }.first()
        l.zip(r).associateTo(allergens) { it }
        unknown -= r
    }

    println(input.map { (it.first - allergens.keys).size }.sum())
    println(allergens.entries.sortedBy { it.value }.map { it.key }.joinToString(","))
}

File: day21/day21.py
import functools
import re

input_ = [(set(m[1].split(' ')), set(m[2].split(', ')))
          for m in [re.match(r'(.+) \(contains (.+)\)', line)
          for line in open("day21.txt")]]

allergens = {}
unknown = set(s for x in input_ for s in x[1])
while unknown:
    for test in unknown:
        l, r = [functools.reduce(lambda a, b: a & b, t)
                for t in zip(*[x for x in input_ if test in x[1]])]
        l -= set(allergens.keys())
        r -= set(allergens.values())
        if len(l) == len(r) and r:
            allergens.update(zip(l, r))
            unknown -= r
            break

clean = lambda s: s - set(allergens.keys())
print(functools.reduce(lambda c, x: c + len(clean(x[0])), input_, 0))
print(','.join(x[0] for x in sorted(allergens.items(), key=lambda x: x[1])))
